### PR TITLE
CM: Allow displaying components in the list view

### PR DESCRIPTION
**Pull request by HichamELBSI**
### What does it do?

Give the possibility to display repeatable components fields in the content manager list view.

Screenshot 2022-03-11 at 14 14 56

### How to test it?

1. Create a collection type
2. Add a repeatable component to this collection type and save
3. Add a single component to this collection type and save
4. Create a new entry and fill several values for the repeatable component
5. Go back to the collection type list view 
6. Enable the repeatable component field to show it in the list columns

Other things to look out for:

- does it work to individually add the columns to the list?
- does it work to add the columns to the list by default (through edit view)?
- does it work with very long content in the main field of each component?
- does it work with different types of main fields?

Refs: https://github.com/strapi/design-system/pull/552

### CODE REVIEW

File: packages/core/admin/admin/src/content-manager/hooks/useFetchContentTypeLayout/utils/formatLayouts.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/hooks/useFetchContentTypeLayout/utils/formatLayouts.js:149]: Can you add some tests?

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/index.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/components/DynamicTable/index.js:98]: You don't need to pass the layout here but what you can do is to change the mainField in the component's layout

```
{
   metadatas: { //..., mainField: { type, name } 
}
```

So you have all the info needed in your RepeatableCell.

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/tests/hasContent.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/tests/hasContent.js:38]: With the comment below the layout should be `{mainField: { type, name: 'name } }`

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js:8]: This is not a truthy statement as the value at the 0 index can be not empty but the mainField could.

```js
return content.every(value => !isEmpty(value)
```

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js:10]: You don't need this part.

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js:14]: `return !isEmpty(content)`

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:53]: With my comment below you'll have to update this part of the code.

```js
const { mainField: { name, type } } = metadatas
```

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:69]:
```js
<CellValue type={type} value={value[0][name]} />
````

**Review by soupette — CHANGES_REQUESTED** [review]

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:57]: 👍

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:78]: Are we sure about the TBD? We are going to merge this PR on master soon.

WDYT about using the same placeholder we use in the LV temporarily? `-`

Comment (soupette) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:78]: Or we can always use the `id`

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js:10]: So it means that my repeatable component does not have the mainField filled it won't be displayed in the LV?

Why not doing the isEmpty on the entire object?

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (HichamELBSI) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:78]: +1 for the id

Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:78]: I asked @maevalienard for an opinion and waiting for a response. I think `-` is too small to be clickable. I was thinking about "items" as fallback (similar to relations), but would like to discuss that with her first.

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js:10]: Because we need to display something in the dropdown, which is (for now) only the mainField of each item. In case no mainField value is set in any of the repeatable components, there is nothing to display ... Do you have something in mind, that we could display instead?

Comment (HichamELBSI) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js:10]: I would go for the ID. This is also an issue on the relation drop down (edit view). Display nothing or a dash without giving any information about that entry is so confusing. WDYT?

Screenshot 2022-02-09 at 13 43 12

File: packages/core/admin/admin/src/content-manager/hooks/useFetchContentTypeLayout/utils/formatLayouts.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/hooks/useFetchContentTypeLayout/utils/formatLayouts.js:149]: Added in 4cfbfcffad0ea70447fbd77621d1aa156e5f03e2

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js]: Added in c98a4ef87679ed83615f03f8179400c9fef29e49. It seems like this is the default everywhere so far, so I guess it makes sense.

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js]: Done in c51772aa2932ebf2cfa47750d042deac35e3cd4f.

@HichamELBSI Is the issue in the relation dropdown related to that PR? If not, could you create an issue instead? 🙏🏽

Comment (HichamELBSI) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js]: I didn't bring this up for you to fix, it was just to show another similar example.

**Comment by gu-stav** [comment]
@HichamELBSI @soupette Thanks for the great reviews. I think everything should be done for now. If you could give it another 👁️ it would be appreciated.

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (maevalienard) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js]: Thanks @gu-stav, let's stick with that and see with the usage :D

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/tests/hasContent.js
Comment (HichamELBSI) [inline packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/tests/hasContent.js:1]: can you rename the file to `hasContent.test.js` please

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/SingleComponent/index.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/SingleComponent/index.js:15]: No, you are right. I copied that over from your component, but it's not needed actually.

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:118]: I've kept your original code here and looked into the `SimpleMenu`: to me it seemed, that because the items have hover/ focus effects, they indicate that they are clickable. In this case the different components aren't. What do you think? We could potentially add a `disabled` prop (naming tbd) for each `SimpleMenuItem` though ...

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:41]: I think I just just kept what was already there and will double-check.

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:16]: I think I just just kept what was already there and will double-check.

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (HichamELBSI) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:118]: Since we are spreading the props, I think you already can pass an `aria-disabled` to the `MenuItem` component 👍🏽

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/SingleComponent/index.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/SingleComponent/index.js]: Done in 124418cee4e58451430f2f602674522663921f86

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js]: Thanks again for the suggestions. I've replaced the custom menu in 3560ceecda6ff931b6692439d4a97f02fea5d640. It makes the whole component much cleaner.

However there is one leftover:

Screenshot from 2022-02-14 09-29-58

Even with `aria-disabled` the `MenuItem` keeps it's hover/ focus states. Is that something I'd need to address in the design-system itself?

**Comment by gu-stav** [comment]
I've found another problem with the current implementation and need to work on this first.

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (HichamELBSI) [inline packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:56]: 👍🏽  Thanks !

Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:56]: @HichamELBSI There is a discussion above ⏫ about disabling the hover effect, that I'd like to have your opinion on :)

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (HichamELBSI) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js]: It should be focusable IMO for the keyboard navigation (+ screen readers announce). If you disable the MenuItem, you'll not be able to scan the MenuItem that's why `aria-diabled` should be enough.

Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js]: @HichamELBSI I agree on focus. The main question is about `hover` since this seems to be misleading to me ...

Comment (HichamELBSI) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js]: I agree on the `hover` 👍🏽  maybe we could remove the hover when aria-disabled is set to `true` 🤔  WDYT?

Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js]: PR for the design-system: https://github.com/strapi/design-system/pull/525

File: packages/core/admin/admin/src/content-manager/pages/ListView/reducer.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/pages/ListView/reducer.js:73]: This should be singular as we retrieve only one attribute.

Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/pages/ListView/reducer.js:73]: No, we retrieve the whole attributes object, which always contains `type`, but sometimes also `default`. I think plural makes sense here.

**Comment by gu-stav** [comment]
This PR currently waits on a new version of the design-system, which will include:

- https://github.com/strapi/design-system/pull/524
- https://github.com/strapi/design-system/pull/525

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:17]: In which case could this be undefined?

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:16]: I don't really get the point of this function could you explain it a bit ?

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js:10]: In this case you will always have an id field so you can use the same statement as below. Or maybe `content && content.length > 0`

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/tests/hasContent.test.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/tests/hasContent.test.js:50]: Components always have an `id` field.

File: packages/core/admin/admin/src/content-manager/hooks/useFetchContentTypeLayout/utils/formatLayouts.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/hooks/useFetchContentTypeLayout/utils/formatLayouts.js:170]: `mainFieldAttribute` it should be singular.

File: packages/core/admin/admin/src/content-manager/pages/ListView/reducer.js
Comment (soupette) [reply packages/core/admin/admin/src/content-manager/pages/ListView/reducer.js:73]: Well it is an attribute definition which always has a `type` and can have validations such as `required,min,max` and also `default` but it is still an attribute definition of the content type.

**Review by soupette — CHANGES_REQUESTED** [review]

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js]: `getMainFieldType` is a leftover and can be removed.

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js]: `getMainFieldType` is a leftover and can be removed.

File: packages/core/admin/admin/src/content-manager/hooks/useFetchContentTypeLayout/utils/formatLayouts.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/hooks/useFetchContentTypeLayout/utils/formatLayouts.js]: I changed that in eaf8baff85135d95d565378903e8b9c802572971. I don't agree, because I still think `value`, `min`, `max` ... are attributes (plural) of the mainField, but you know the data-structure better than me and I think it's not worth to discuss that any further. 👍🏼

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/tests/hasContent.test.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/tests/hasContent.test.js]: Fixed in 44921d68c2f3b94d626f91dc769aa255481dbd74

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/utils/hasContent.js]: Fixed in 44921d68c2f3b94d626f91dc769aa255481dbd74

**Review by soupette — APPROVED** [review]

File: packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js
Comment (soupette) [inline packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:18]: Why do we need this? Is there any other way rather than using negative margins with hardcoded values?

cc/ @HichamELBSI

Comment (HichamELBSI) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:18]: @gu-stav Can't we reset the `padding` instead of negative margin as we'd said?

Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:18]: @soupette I've discussed this with Hicham option 2.

Here are the cases:

1. `SimpleMenu` without reducing padding-left

It looks really off.

Screenshot 2022-02-23 at 17 02 49

2. `SimpleMenu` with reduced padding

Screenshot 2022-02-23 at 17 03 10

3. `SimpleMenu` with reduced padding and negative margin

Screenshot 2022-02-23 at 17 03 57

That's being said: I'm not a fan of negative margins either, but I think we have to properly align the menu somehow. Reducing the padding already works, but it looks still a bit off.

Also pinging @maevalienard here.

Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:18]: Another idea: we could create, similar to button, a `size="S|M"` variant. In "s" we then could reduce the padding to a minimum ... What do you think about it?

Comment (HichamELBSI) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:18]: Great idea! Since we are using this component in several context, it should be customizable

Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:18]: I've opened a PR here: https://github.com/strapi/design-system/pull/552 and leave this discussion open until merged.

The question if we agree on using a negative margin here still needs to be discussed.

Comment (soupette) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:18]: Well if you really need the negative margin I'm fine with it however can you use the sizes from the DS instead of hardcoding `6px` so it still consistent?

Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/RepeatableComponent/index.js:18]: @soupette I'll do my best :)

**Comment by gu-stav** [comment]
@soupette @maevalienard No that the DS released the small variant for `SimpleMenu`s, I refactored the PR to make use of it and updated the Screenshot in the PR description.

**Review by HichamELBSI — COMMENTED** [review]
LGTM

**Review by soupette — APPROVED** [review]

**Review by soupette — APPROVED** [review]

**Review by MarionLemaire — APPROVED** [review]
Tested on Getstarted using Chrome : 

- Main Case OK
- Edge cases : 
- individually add the columns to the list : OK
- add the columns to the list by default (through edit view) : OK
- with very long content in the main field of each component : OK 
- with different types of main fields? OK (works with text field, phone field, ID by default >> has to be checked with a complete list of main fields later ?

**Pull request merged**
